### ISSUE TITLE
segfault in perftools::gputools::StreamExecutor::DeviceMemoryUsage - on a busy gpu

### ISSUE THREAD

**Issue by MInner** [tensorflow/tensorflow#2840]
I'm using tensorflow 0.9.0rc0 with cuda 7.5 on Tesla K40c .

The GPU I'm specifying via `device_id` and

```
    gpu_options = tf.GPUOptions(per_process_gpu_memory_fraction=0.7, allow_growth=True)
    sess_cfg = tf.ConfigProto(allow_soft_placement=True, gpu_options=gpu_options)
```

is running under a heavy load now (multiple instances of tf are working with it in parallel, so it has 70-80% gpu voltage), but still has ~20-30% free memory, so I wanted to run a one more tiny script.

If I run the script under `gdb`, I get:

```
Program received signal SIGSEGV, Segmentation fault.
0x00007fffd51237a1 in perftools::gputools::StreamExecutor::DeviceMemoryUsage(long long*, long long*) const ()
   from /data2/users/usman/anaconda2/envs/tfpy3/lib/python3.5/site-packages/tensorflow/python/_pywrap_tensorflow.so
```

is it indeed "out of resources" or there's something wrong with the installation? Shouldn't it somehow signalize about those problems in user code, rather than just segfaulting?

**Comment by zheng-xq** [comment]
@MInner, the error comes from stream-executor. I will try to loop in some developers to help there. 

Something seems wrong here. If you don't run this script under "gdb", what is the output? Could you post the entire log somewhere, and post the link here?

**Comment by MInner** [comment]
Hi, @zheng-xq . If I run it without `gdb`, nothing interesting happens (no error messages or whatever, just the ordinary debug output regarding cuda .so files and etc.) until I try to execute the graph, and then it just segfaults silently (before any "allocation attempt" messages pop up). The very same script works just fine, if the GPU is not under heavy load; but when it was (under load), it failed with 100% chance. What kind of log should I share? The `dgb` output is just "[Thread XXXX]" and then the segfault message I posted above. I also can't see anything in `/var/crash/` (that is server Ubuntu 14.04.02, so apport should have posted something there, but it didn't, hmm)

There is a _very_ little change that this is a gpu driver problem (and it is the thing that causes segfault under heavy load), because I remember that we used to experience some issues with gpu drives at some point in the past.

**Comment by zheng-xq** [comment]
Swapping my action with Jason, who owns stream-executor. Feel free to assign back to me when we feel TensorFlow is at fault here. 

The segfault is suspicious. Well behaving TensorFlow or stream-executor code should check for error conditions and quit with a clear error message.

**Comment by zheng-xq** [comment]
Really adding Jason. +@henline

**Comment by henline** [comment]
Hi @MInner. Thanks for your report.

I think it would be useful for me to see the full stdout and stderr for the run that causes the segfault. That should include all the ordinary debug output you mentioned (loading cuda .so files, etc.).

I'd also like to see the full backtrace when you run it under gdb. You should be able to get that by using the gdb 'backtrace' command.

**Comment by MInner** [comment]
Sorry for not answering that long, I needed to wait for a heavy cluster load to reproduce this error

here's the `gdb` output and the backtrace:
https://gist.github.com/MInner/dff6d209948d95cf4da9245dbbab18e8 .

And yes, I totally forgot about the user-level warning regarding allocating a huge dense matrix - it comes from the modified code snippet `last_relevant()` from #206 that extracts "last relevant output" of dynamically unrolled rnn with different input lengths .

**Comment by mhng1580** [comment]
Hi there,

I think I get the similar error as @MInner
Also, this happens on only when I run on a 6G GPU. The program works normally on a 12G GPU card. 
It would be great if there's any followup. 
Here's my debug output and backtrace
```
Starting program: /data/sls/scratch/wnhsu/pyenv_tf/r08/bin/python dnn_train_galebc2_debug.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcurand.so locally
[New Thread 0x7fffd82e6700 (LWP 9747)]
[New Thread 0x7fffd7ae5700 (LWP 9748)]
[New Thread 0x7fffd52e4700 (LWP 9749)]
[New Thread 0x7fffd2ae3700 (LWP 9750)]
[New Thread 0x7fffd02e2700 (LWP 9751)]
[New Thread 0x7fffcdae1700 (LWP 9752)]
[New Thread 0x7fffcb2e0700 (LWP 9753)]
[Thread 0x7fffd7ae5700 (LWP 9748) exited]
[Thread 0x7fffd52e4700 (LWP 9749) exited]
[Thread 0x7fffd82e6700 (LWP 9747) exited]
[Thread 0x7fffcdae1700 (LWP 9752) exited]
[Thread 0x7fffd02e2700 (LWP 9751) exited]
[Thread 0x7fffd2ae3700 (LWP 9750) exited]
[Thread 0x7fffcb2e0700 (LWP 9753) exited]
building DNN model with 5 hidden layers of 1024 hidden units
building train_op with constant learning rate 0.001
Loading data into memory...
ali-to-post ark:- ark:- 
ali-to-pdf /data/sls/scratch/wnhsu/gale_mandarin_bc_phase2/exp/tri3b_dnn_2048x5_ali/final.mdl 'ark:gunzip -c /data/sls/scratch/wnhsu/gale_mandarin_bc_phase2/exp/tri3b_dnn_2048x5_ali/ali.*.gz |' ark:- 
LOG (ali-to-pdf:main():ali-to-pdf.cc:68) Converted 76184 alignments to pdf sequences.
LOG (ali-to-post:main():ali-to-post.cc:65) Converted 76184 alignments.
   7804 utterances loaded...
   avg-sequence-len = 608
[New Thread 0x7fffcb2e0700 (LWP 9764)]
[New Thread 0x7fffcdae1700 (LWP 9765)]
[New Thread 0x7fffd02e2700 (LWP 9766)]
[New Thread 0x7fffd2ae3700 (LWP 9767)]
[New Thread 0x7fffa9dda700 (LWP 9768)]
[New Thread 0x7fffa95d9700 (LWP 9769)]
[New Thread 0x7fffa8dd8700 (LWP 9770)]
[New Thread 0x7fffa5484700 (LWP 9771)]
[New Thread 0x7fffa4c83700 (LWP 9772)]
[Thread 0x7fffa4c83700 (LWP 9772) exited]

Program received signal SIGSEGV, Segmentation fault.
0x00007fffed6730b8 in perftools::gputools::StreamExecutor::DeviceMemoryUsage(long long*, long long*) const ()
   from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
(gdb) backtrace
#0  0x00007fffed6730b8 in perftools::gputools::StreamExecutor::DeviceMemoryUsage(long long*, long long*) const ()
   from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
#1  0x00007fffed40e258 in tensorflow::GPUMachineManager() () from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
#2  0x00007fffed40c4d0 in tensorflow::BaseGPUDeviceFactory::GetValidDeviceIds(std::vector<int, std::allocator<int> >*) ()
   from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
#3  0x00007fffed40cd50 in tensorflow::BaseGPUDeviceFactory::CreateDevices(tensorflow::SessionOptions const&, std::string const&, std::vector<tensorflow::Device*, std::allocator<tensorflow::Device*> >*) ()
   from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
#4  0x00007fffed5e50d6 in tensorflow::DeviceFactory::AddDevices(tensorflow::SessionOptions const&, std::string const&, std::vector<tensorflow::Device*, std::allocator<tensorflow::Device*> >*) ()
   from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
#5  0x00007fffed3d0b71 in tensorflow::DirectSessionFactory::NewSession(tensorflow::SessionOptions const&) ()
   from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
#6  0x00007fffed6076d7 in tensorflow::NewSession(tensorflow::SessionOptions const&, tensorflow::Session**) ()
   from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
#7  0x00007fffed5d60e1 in TF_NewSession () from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
#8  0x00007fffec839c3b in _wrap_TF_NewSession () from /data/sls/scratch/wnhsu/pyenv_tf/r08/local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow.so
#9  0x000000000049968d in PyEval_EvalFrameEx ()
#10 0x00000000004a090c in PyEval_EvalCodeEx ()
#11 0x0000000000499a52 in PyEval_EvalFrameEx ()
#12 0x00000000004a1c9a in ?? ()
#13 0x00000000004dfe94 in ?? ()
#14 0x0000000000505f96 in PyObject_Call ()
#15 0x00000000004de41a in ?? ()
#16 0x00000000005039eb in ?? ()
#17 0x0000000000499be5 in PyEval_EvalFrameEx ()
#18 0x00000000004a090c in PyEval_EvalCodeEx ()
#19 0x000000000049ab45 in PyEval_EvalFrameEx ()
#20 0x00000000004a1634 in ?? ()
#21 0x000000000044e4a5 in PyRun_FileExFlags ()
#22 0x000000000044ec9f in PyRun_SimpleFileExFlags ()
#23 0x000000000044f904 in Py_Main ()
#24 0x00007ffff7818f45 in __libc_start_main (main=0x44f9c2 <main>, argc=2, argv=0x7fffffffd3b8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd3a8)
    at libc-start.c:287
#25 0x0000000000578c4e in _start ()
```

**Comment by henline** [comment]
Looking through the code, I didn't see any obvious culprit for this. I'm having a hard time reproducing it on my own.

@MInner or @mhng1580 do you happen to have a simple reproducer for this which includes all the data and python commands needed to trigger the segfault?

**Comment by gnujimmik** [comment]
I have the similar segfault when I run cifar10_eval.py. My env is python3.5, CUDA7.5 with cuDNN 5, tensorflow 0.9, ubuntu 14.04. When I ran the code with tf 0.8, it was fine. 

When I run gdb with the code. I've got the segfault as below:
Program received signal SIGSEGV, Segmentation fault.
0x00007fffebbc0011 in perftools::gputools::StreamExecutor::DeviceMemoryUsage(long long*, long long*) const ()
   from /opt/conda/lib/python3.5/site-packages/tensorflow/python/_pywrap_tensorflow.so

Anyone knows how to fix this?

**Comment by hbfs** [comment]
This definitely looks like its related to the amount of free memory available.

This is tested on: 
```
OS Version:            Mac OS X 10.11.6 (15G31)
Kernel Version: Darwin 15.6.0
Graphics: NVIDIA GeForce GTX 960, NVIDIA GeForce GTX 960, PCIe 2GB
tensorflow version: 0.9.0 (8a4f6abb395b3f1bca732797068021c786c1ec76)
```
I'm using a monitor thats 4096x2160 with Billions of Colors and 60 Hz. This is relevant because it directly impacts the amount of free memory available on the GPU. Also just as important is how many windows are displayed on the current desktop.

1) With many windows, Chrome tabs and viewing a Youtube video:
```
$ python test.py
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.7.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.7.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.7.5.dylib locally
[1]    54037 segmentation fault  python test.py
```

2) Same desktop, closing the youtube video and some chrome windows:
```
$ python test.py                                                                                                                            139 ↵
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.7.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.7.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.1.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.7.5.dylib locally
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:892] OS X does not support NUMA - returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties:
name: GeForce GTX 960
major: 5 minor: 2 memoryClockRate (GHz) 1.367
pciBusID 0000:04:00.0
Total memory: 2.00GiB
Free memory: 80.17MiB
I tensorflow/core/common_runtime/gpu/gpu_init.cc:126] DMA: 0
I tensorflow/core/common_runtime/gpu/gpu_init.cc:136] 0:   Y
I tensorflow/core/common_runtime/gpu/gpu_device.cc:839] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 960, pci bus id: 0000:04:00.0)
Device mapping:
/job:localhost/replica:0/task:0/gpu:0 -> device: 0, name: GeForce GTX 960, pci bus id: 0000:04:00.0
I tensorflow/core/common_runtime/direct_session.cc:175] Device mapping:
/job:localhost/replica:0/task:0/gpu:0 -> device: 0, name: GeForce GTX 960, pci bus id: 0000:04:00.0

MatMul: /job:localhost/replica:0/task:0/gpu:0
I tensorflow/core/common_runtime/simple_placer.cc:818] MatMul: /job:localhost/replica:0/task:0/gpu:0
b: /job:localhost/replica:0/task:0/gpu:0
I tensorflow/core/common_runtime/simple_placer.cc:818] b: /job:localhost/replica:0/task:0/gpu:0
a: /job:localhost/replica:0/task:0/gpu:0
I tensorflow/core/common_runtime/simple_placer.cc:818] a: /job:localhost/replica:0/task:0/gpu:0
E tensorflow/stream_executor/cuda/cuda_driver.cc:965] failed to allocate 80.17M (84066304 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY
[[ 22.  28.]
 [ 49.  64.]]
```

3) Switching to a desktop with no windows open:
```
$ python test.py
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.7.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.7.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.1.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.7.5.dylib locally
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:892] OS X does not support NUMA - returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties:
name: GeForce GTX 960
major: 5 minor: 2 memoryClockRate (GHz) 1.367
pciBusID 0000:04:00.0
Total memory: 2.00GiB
Free memory: 1.05GiB
I tensorflow/core/common_runtime/gpu/gpu_init.cc:126] DMA: 0
I tensorflow/core/common_runtime/gpu/gpu_init.cc:136] 0:   Y
I tensorflow/core/common_runtime/gpu/gpu_device.cc:839] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 960, pci bus id: 0000:04:00.0)
Device mapping:
/job:localhost/replica:0/task:0/gpu:0 -> device: 0, name: GeForce GTX 960, pci bus id: 0000:04:00.0
I tensorflow/core/common_runtime/direct_session.cc:175] Device mapping:
/job:localhost/replica:0/task:0/gpu:0 -> device: 0, name: GeForce GTX 960, pci bus id: 0000:04:00.0

MatMul: /job:localhost/replica:0/task:0/gpu:0
I tensorflow/core/common_runtime/simple_placer.cc:818] MatMul: /job:localhost/replica:0/task:0/gpu:0
b: /job:localhost/replica:0/task:0/gpu:0
I tensorflow/core/common_runtime/simple_placer.cc:818] b: /job:localhost/replica:0/task:0/gpu:0
a: /job:localhost/replica:0/task:0/gpu:0
I tensorflow/core/common_runtime/simple_placer.cc:818] a: /job:localhost/replica:0/task:0/gpu:0
[[ 22.  28.]
 [ 49.  64.]]
```

**Comment by drpngx** [comment]
Closing since this is based on old code that changed substantially in the last six months. Feel free to open a new issue if the problem persists with new code.
